src/Compilers/Core/MSBuildTask/DotnetHost.cs

```diff
@@ -0,0 +1,61 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.CodeAnalysis.BuildTasks
+{
+    internal class DotnetHost
+    {
+        public string ToolName { get; }
+
+        public string PathToToolOpt { get; }
+
+        public string CommandLineArgs { get; }
+
+        public DotnetHost(string toolName, string pathToTool, string commandLineArgs)
+        {
+            ToolName = toolName;
+            PathToToolOpt = pathToTool;
+            CommandLineArgs = commandLineArgs;
+        }
+
+        public DotnetHost(string toolNameWithoutExtension, string commandLineArgs)
+        {
+            string pathToTool;
+            string toolName;
+            if (IsCliHost(out string pathToDotnet))
+            {
+                toolName = $"{toolNameWithoutExtension}.dll";
+                pathToTool = Utilities.GenerateFullPathToTool(toolName);
```

Comment:
Probably, especially if we pass in `DOTNET_HOST_PATH` as a task variable. The reason I made this class because the three properties in this class need to be created in tandem (as they have intertwined logic), so when any one of the three are requested, all three of them are computed (by calling the `DotnetHost` property on ManagedCompiler).

(Passing the env var as a property might simplify it to the point that it's easy to just re-do the work on any one of the three property requests - it'd probably just be an is-null check.)